agent/submatview/store.go

```diff
@@ -0,0 +1,237 @@
+	expiryHeap *ttlcache.ExpiryHeap
+
+	// idleTTL is the duration of time an entry should remain in the Store after the
+	// last request for that entry has been terminated. It is a field on the struct
+	// so that it can be patched in tests without need a lock.
+	idleTTL time.Duration
+}
+
+type entry struct {
+	materializer *Materializer
+	expiry       *ttlcache.Entry
+	stop         func()
+	// requests is the count of active requests using this entry. This entry will
+	// remain in the store as long as this count remains > 0.
+	requests int
+}
+
+// NewStore creates and returns a Store that is ready for use. The caller must
+// call Store.Run (likely in a separate goroutine) to start the expiration loop.
+func NewStore(logger hclog.Logger) *Store {
+	return &Store{
+		logger:     logger,
+		byKey:      make(map[string]entry),
+		expiryHeap: ttlcache.NewExpiryHeap(),
+		idleTTL:    20 * time.Minute,
+	}
+}
+
+// Run the expiration loop until the context is cancelled.
+func (s *Store) Run(ctx context.Context) {
+	for {
+		s.lock.RLock()
+		timer := s.expiryHeap.Next()
+		s.lock.RUnlock()
+
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+			return
+
+		// the first item in the heap has changed, restart the timer with the
+		// new TTL.
+		case <-s.expiryHeap.NotifyCh:
+			timer.Stop()
+			continue
+
+		// the TTL for the first item has been reached, attempt an expiration.
+		case <-timer.Wait():
+			s.lock.Lock()
+
+			he := timer.Entry
+			s.expiryHeap.Remove(he.Index())
+
+			e := s.byKey[he.Key()]
+
+			// Only stop the materializer if there are no active requests.
+			if e.requests == 0 {
+				e.stop()
+				delete(s.byKey, he.Key())
+			}
+
+			s.lock.Unlock()
+		}
+	}
+}
+
+// Request is used to request data from the Store.
+// Note that cache.Request is required, but some of the fields cache.RequestInfo
+// fields are ignored (ex: MaxAge, and MustRevalidate).
+type Request interface {
+	cache.Request
+	// NewMaterializer will be called if there is no active materializer to fulfil
+	// the request. It should return a Materializer appropriate for streaming
+	// data to fulfil this request.
+	NewMaterializer() (*Materializer, error)
+	// Type should return a string which uniquely identifies this type of request.
+	// The returned value is used as the prefix of the key used to index
+	// entries in the Store.
+	Type() string
+}
+
+// Get a value from the store, blocking if the store has not yet seen the
+// req.Index value.
+// See agent/cache.Cache.Get for complete documentation.
+func (s *Store) Get(ctx context.Context, req Request) (Result, error) {
+	info := req.CacheInfo()
+	key, materializer, err := s.readEntry(req)
+	if err != nil {
+		return Result{}, err
+	}
+	defer s.releaseEntry(key)
+
+	ctx, cancel := context.WithTimeout(ctx, info.Timeout)
+	defer cancel()
+
+	result, err := materializer.getFromView(ctx, info.MinIndex)
+	// context.DeadlineExceeded is translated to nil to match the behaviour of
+	// agent/cache.Cache.Get.
```

Comment:
I wonder if this could mask underlying network timeouts too?

I assume gRPC client code also uses contexts for things like connection timeout, if it did and there was an actual error like unable to connect to servers which caused a timeout, could we mistakenly mask that error?

I wonder if there is a way to be sure this is the timeout based on the user's requested blocking timeout vs a downstream network timeout that's propagated directly back through the calls? Right now this might never be possible but it makes me wonder if this could ever result in a subtle bug later?

Reply:
From what I can tell request timeouts in gRPC are always set by the context, so there's no difference. Connect timeouts are reported using this error: https://github.com/grpc/grpc-go/blob/24d03d9f769106b3c96b4145244ce682999d3d88/internal/transport/transport.go#L713, so would not be `context.DeadlineExceeded`.

I believe this is safe. I do think it is unfortunate, and it would be better to report this timeout to the user, but agent/cache masks these kinds of errors, so I thought for now it would be better to match the behaviour.